### PR TITLE
Make it possible to test runners that don't support all metrics

### PR DESCRIPTION
**Pull request by dhalperi**
Runners may only partially support metrics. This partial support runs along axes:
- attempted & committed
- counters & distributions & gauges & ...

Prior to this PR, we have categories for the first axis, but not the second.
This means that a runner that supports only part of the second axis has to blacklist tests
on the first axis. Adding categories for both axes lets runners build a matrix of supported
features.

I also renamed the existing categories so that they are grouped alphabetically. Happy to
undo if this is not a good change.

This also lets the DataflowRunner run more tests by accurately identifying its test
matrix.

### CODE REVIEW

**Comment by dhalperi** [comment]
R: @aviemzur 

I know you have designs on making the second axis irrelevant, but as it's not yet irrelevant I'd like to make this change now to improve test coverage of existing runners.

**Comment by aviemzur** [comment]
Run Dataflow ValidatesRunner

**Comment by aviemzur** [comment]
Looks like the 2 main tests in `MetricsTest` still basically use all of the metrics types. So they are still not runnable by Dataflow runner.
Perhaps split these tests or add more tests, each using a specific metric type?

**Comment by aviemzur** [comment]
Regarding the category names:

What was the reason for the changes:
`UsesAttemptedMetrics` -> `UsesMetricsAttempted`
`UsesCommittedMetrics` -> `UsesMetricsCommitted`

And the new categories:
`UsesMetricsCounter`
`UsesMetricsDistribution`
`UsesMetricsGauge`

Is the format of `UsesMetrics*` for searchability/completion reasons?
I believe modern IDEs have solved this issue and we should opt for more human comprehensible names instead.
This would be more consistent with other category names which also opted for more human comprehensible names. For example: `UsesTimersInParDo` and `UsesSplittableParDoWithWindowedSideInputs`.

**Comment by dhalperi** [comment]
run Dataflow ValidatesRunner

**Comment by dhalperi** [comment]
run Dataflow ValidatesRunner

**Comment by dhalperi** [comment]
retest this please

**Comment by dhalperi** [comment]
jenkins appears to be kinda hosed, but @aviemzur please take another look.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java [hunk 1]
```diff
@@ -121,62 +123,89 @@ public void counterToCell() {
...
     MetricQueryResults metrics = result.metrics().queryMetrics(MetricsFilter.builder()
         .addNameFilter(MetricNameFilter.inNamespace(MetricsTest.class))
         .build());
 
-    assertThat(metrics.counters(), hasItem(
-        committedMetricsResult(NAMESPACE, "count", "MyStep1", 3L)));
-    assertThat(metrics.distributions(), hasItem(
-        committedMetricsResult(NAMESPACE, "input", "MyStep1",
-            DistributionResult.create(26L, 3L, 5L, 13L))));
+    testAllMetrics(metrics, true);
```
Comment (aviemzur) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java]: Rename suggestion: rename `test*` methods which aren't in themselves tests to `assert*`.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java [hunk 2]
```diff
@@ -121,62 +123,89 @@ public void counterToCell() {
...
-            DistributionResult.create(52L, 6L, 5L, 13L))));
-    assertThat(metrics.gauges(), hasItem(
-        committedMetricsResult(NAMESPACE, "my-gauge", "MyStep2",
-            GaugeResult.create(12L, Instant.now()))));
+  @Category({ValidatesRunner.class, UsesAttemptedMetrics.class, UsesCounterMetrics.class,
+      UsesDistributionMetrics.class, UsesGaugeMetrics.class})
+  @Test
+  public void testAllAttemptedMetrics() {
+    PipelineResult result = runPipelineWithMetrics();
+    MetricQueryResults metrics = result.metrics().queryMetrics(MetricsFilter.builder()
```
Comment (aviemzur) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java]: This piece of code seems appears 8 times in `MetricsTest`:
```java
MetricQueryResults metrics = result.metrics().queryMetrics(MetricsFilter.builder()
    .addNameFilter(MetricNameFilter.inNamespace(MetricsTest.class))
    .build());
```
Perhaps it is worth a helper method such as 
```java
MetricQueryResults queryMetrics(PipelineResult result) { ... }
```

File: sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java (hunk 1 above)
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java (hunk 2 above)
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java]: done

**Pull request closed without merging**